### PR TITLE
Add regression test for bug#11106

### PR DESCRIPTION
**Pull request by NthPortal**
Resubmit of #7472

### CODE REVIEW

**Comment by som-snytt** [comment]
More minimal, or more visual:
```
scala 2.13.0-M4> Iterator.iterate(0) { _ => print("X") ; 0 }.take(3).toList
XXres0: List[Int] = List(0, 0, 0)

scala 2.13.0-M4> Iterator.iterate(0) { _ => print("X") ; 0 }.withFilter(_ != -1).take(3).toList
XXXres1: List[Int] = List(0, 0, 0)
```
```
scala 2.13.0-M5> Iterator.iterate(0) { _ => print("X") ; 0 }.withFilter(_ != -1).take(3).toList
XXres0: List[Int] = List(0, 0, 0)
```
Diagnosed [here](https://github.com/scala/scala/pull/6679#issuecomment-392084461).  The bug was introduced early in strawman; the fix was to revert to 2.12 use of slice. Underlying `hasNext` was called an extra time after take count was zero. That's why the behavior changes with filter.

One might have assumed that 2.13 collections would require automated testing to shake out these inevitable bugs, but such a test rig never materialized.

File: test/junit/scala/collection/IteratorTest.scala [hunk 1]
```diff
@@ -600,4 +600,15 @@ class IteratorTest {
...
   }
+
+  @Test
+  def t11106(): Unit = {
+    var i = 0
+    Iterator.continually(0)
+      .map(_ => {i += 1; i})
+      .withFilter(_ < 10)
+      .take(3)
+      .foreach(_ => ())
```
Comment (som-snytt) [inline test/junit/scala/collection/IteratorTest.scala:611]: For me, it's easier to see with 
```
scala 2.13.0-M4> var i = 0; Iterator.continually { i += 1; i }.withFilter(_ => true).take(3).toList
i: Int = 4
res0: List[Int] = List(1, 2, 3)
```
where both `i` and the length should be 3.

**Review by som-snytt — APPROVED** [review]
Suggested an improvement? but nbd.

**Comment by som-snytt** [comment]
Now I know [what I was doing](https://github.com/scala/scala/commit/bc1b82e37d59df3a758371c0848871311aa67056) on Halloween four years ago.

The comment notes the other good reason for slicing:
```
scala 2.13.0-M4> Iterator.continually { ??? }.take(1).take(1).take(1).take(1)
res0: Iterator[Nothing] = non-empty iterator

scala 2.13.0-M4> .toList
scala.NotImplementedError: an implementation is missing
  at scala.Predef$.$qmark$qmark$qmark(Predef.scala:340)
  at .$anonfun$res0$1(<console>:1)
  at scala.collection.Iterator$$anon$25.next(Iterator.scala:972)
  at scala.collection.Iterator$$anon$10.next(Iterator.scala:522)
  at scala.collection.Iterator$$anon$10.next(Iterator.scala:522)
  at scala.collection.Iterator$$anon$10.next(Iterator.scala:522)
  at scala.collection.Iterator$$anon$10.next(Iterator.scala:522)
  at scala.collection.mutable.Growable.addAll(Growable.scala:50)
  at scala.collection.mutable.Growable.addAll$(Growable.scala:47)
  at scala.collection.mutable.AbstractBuffer.addAll(Buffer.scala:174)
  at scala.collection.mutable.ListBuffer$.from(ListBuffer.scala:289)
  at scala.collection.immutable.List$.from(List.scala:508)
  at scala.collection.IterableOnceOps.toList(IterableOnce.scala:1053)
  at scala.collection.IterableOnceOps.toList$(IterableOnce.scala:1053)
  at scala.collection.AbstractIterator.toList(Iterator.scala:1096)
  ... 30 elided
```

**Comment by som-snytt** [comment]
This bug must have a famous name in the literature, because apparently it happens [all the time](https://github.com/scala/scala/commit/71cfb57499bda80f5fb0a810f37d0f9c54e1afd3).

File: test/junit/scala/collection/IteratorTest.scala (hunk 1 above)
Comment (NthPortal) [reply test/junit/scala/collection/IteratorTest.scala:611]: I have no strong preference; I just copied from the bug ticket.

What you suggest seems fine to me. Feel free to make the change and force-push if you want - I won't be able to do get to it for a few days at least.

**Comment by lrytz** [comment]
(i'm currently in the getting-prs-merged mode)

**Pull request merged**